### ISSUE TITLE
Rasa-X deployment gets updated, although nothing has changed

### ISSUE THREAD

**Issue by loomsen** [RasaHQ/rasa-x-helm#69]
Hi.

I'm trying to understand the reasoning behind the Deployment Strategy type of `Rereate` for rasa-x. As far as I understand, rasa-x is just the user interface. Why do we have to `Recreate` if it's just the UI? It integrates with git anyway to pull the latest changes, no?

Also what's the reasoning behind the checksum parameters? This will almost always trigger a redeploy. It breaks our workflow, and more importantly, the workflow of our content creator colleagues, who then ask us why "it's broken all the time".

Thank you.

**Comment by sara-tagger** [comment]
Thanks for the issue, @b-quachtran will get back to you about it soon!

###### You may find help in the [docs](https://rasa.com/docs/) and the [forum,](https://forum.rasa.com/) too 🤗

**Comment by tczekajlo** [comment]
@loomsen 👋  The Rasa X or the rasa-x deployment it's not only the UI, but it's also the Rasa X server that servers API (https://rasa.com/docs/rasa-x/api/rasa-x-http-api/). For now, we use a persistent volume in the rasa-x deployment, that's why we use the `Recreate` strategy. 

The reasoning behind the checksum parameters is to be sure that deployment is up-to-date if a configuration for it changes.  If checksum changed then some values had to change as well. You can use the helm diff plugin to check what changes are going to be applied, e.g.

```
helm diff upgrade --version=1.6.0 -f values-test.yaml rasa-x rasa-x/rasa-x
```

**Comment by loomsen** [comment]
@tczekajlo Hi and thank you for your answer.

I find it a little hard to follow to be honest. If the configuration changes, it's k8s' job to figure that out and trigger a rollout, no? 
Now, when our content creators create stories and push them to git, nothing has actually changed in the rasa-x UI nor the API server. It's just user generated data. But still this will trigger a redeploy if we run `helm upgrade -i`. 

I think it shouldn't. It creates unneccesary downtime of up to one minutes, where, in a cloud environment, the pods could be killed at any time and migrated to another node or so, which would add to this downtime.

But that's just my 2ct, thank you for taking the time to answer my questions :)

**Comment by tczekajlo** [comment]
> @tczekajlo Hi and thank you for your answer.
> 
> I find it a little hard to follow to be honest. If the configuration changes, it's k8s' job to figure that out and trigger a rollout, no?

Yes, you're right 😄  but if the deployment spec itself didn't change the application keeps running, e.g if a config map has changed, k8s doesn't rollout upgrade.  That's why the `checksum` annotation is used.

> Now, when our content creators create stories and push them to git, nothing has actually changed in the rasa-x UI nor the API server. It's just user generated data. But still this will trigger a redeploy if we run `helm upgrade -i`.

Changes pushed to a git repository that is used along with Rasa X shouldn't trigger an upgrade. Please make sure that the values used during the helm upgrade haven't changed, that version of the helm chart is pinned.

> 
> I think it shouldn't. It creates unneccesary downtime of up to one minutes, where, in a cloud environment, the pods could be killed at any time and migrated to another node or so, which would add to this downtime.
> 
> But that's just my 2ct, thank you for taking the time to answer my questions :)

**Comment by loomsen** [comment]
Hi @tczekajlo! Sorry for the late reply, got dragged into another project.

> Changes pushed to a git repository that is used along with Rasa X shouldn't trigger an upgrade. Please make sure that the values used during the helm upgrade haven't changed, that version of the helm chart is pinned.

In theory, this is a good idea. However, if you look at [this line in the rasa-x chart](https://github.com/RasaHQ/rasa-x-helm/blob/f24cd3c82f4771753cd43c9db4c03cdb78fff472/charts/rasa-x/templates/rasa-x-deployment.yaml#L23), it becomes clear why rasa-x is redeployed, even though nothing concerning rasa-x changed. 
Our developers are working on actions, so the image tag for rasa-app changes with every commit in development (the tag actually being the commit SHA). I think this is a pretty common pattern.

Thus, changes to rasa-app will trigger a redeploy of rasa-x, which seems unreasonable to me. But maybe I still don't understand all the hot wiring and why this is needed. It's definitely something that's breaking some (as I suppose pretty common) setups and workflows.

**Comment by tczekajlo** [comment]
> Hi @tczekajlo! Sorry for the late reply, got dragged into another project.
> 
> > Changes pushed to a git repository that is used along with Rasa X shouldn't trigger an upgrade. Please make sure that the values used during the helm upgrade haven't changed, that version of the helm chart is pinned.
> 
> In theory, this is a good idea. However, if you look at [this line in the rasa-x chart](https://github.com/RasaHQ/rasa-x-helm/blob/f24cd3c82f4771753cd43c9db4c03cdb78fff472/charts/rasa-x/templates/rasa-x-deployment.yaml#L23), it becomes clear why rasa-x is redeployed, even though nothing concerning rasa-x changed.
> Our developers are working on actions, so the image tag for rasa-app changes with every commit in development (the tag actually being the commit SHA). I think this is a pretty common pattern.
> 
> Thus, changes to rasa-app will trigger a redeploy of rasa-x, which seems unreasonable to me. But maybe I still don't understand all the hot wiring and why this is needed. It's definitely something that's breaking some (as I suppose pretty common) setups and workflows.

If I got all parts of the workflow that you have it should look like this:

Changes in actions -> Build a new image with actions -> Deploy a new image, the only thing that is changed in the helm values is a tag for the rasa-app image(?)

Did I miss something?

**Comment by loomsen** [comment]
Hi @tczekajlo 

Yes, you got that exactly right. 

Here's the command I use in the CI/CD pipeline. `$CI_COMMIT_REF_SLUG`, `$deployment_url` and the values don't change. Only things that change are rasa.tag and app.tag.
```
 helm upgrade $CI_COMMIT_REF_SLUG rasa-x/rasa-x
      -f deployment/rasa-x/values.yaml
      -i --namespace $CI_COMMIT_REF_SLUG
      --create-namespace
      --set ingress.hosts[0].host=$deployment_url
      --set ingress.tls[0].hosts[0]=$deployment_url
      --set ingress.tls[0].secretName=$CI_COMMIT_REF_SLUG-tls
      --set rasa.tag=$CI_COMMIT_SHA
      --set app.tag=$CI_COMMIT_SHA
      --wait
      --timeout 1200s
```

The latest installations were using Chart version 1.6.1, but we're not setting a specific chart version, so `latest`.
